src/osd/OSD.cc

```diff
@@ -10547,6 +10557,7 @@ void OSDShard::prime_merges(const OSDMapRef& as_of_osdmap,
       PGRef pg = osd->handle_pg_create_info(shard_osdmap, &cinfo);
       _attach_pg(r.first->second.get(), pg.get());
       _wake_pg_slot(pgid, slot);
+      osd->_prime_pg_splits_and_merges(pg);
```

Comment:
Is it possible the split children might require the same shard_lock during priming? In that case we might need to promote the shard_lock to be a recursive one...

Reply:
probably?  maybe that's what happened.. I ran this through qa way back when and had lots of failures.  I haven't gotten back to it since the bug is pretty rare.